### DISCUSSION TITLE
Possible to have dataclass-like parameters?

### DISCUSSION THREAD

**Discussion by blakeNaccarato** [DanCardin/cappa#137] [Q&A] [answered]
If I have a `cappa.command`-decorated class (e.g. `Binarize`	below), which I intend to be a terminating command with no subcommands, but one of its fields/attributes happens to be dataclass-like, I believe Cappa assumes it to be an incomplete continuation of CLI structure, e.g. an unsatisfied dependency in the CLI structure.

Dataclass-like structures are special in that they are promoted to command/subcommand status, but is there a way to tell Cappa "treat this not as a subcommand/unsatisfied dep, but rather treat it as a dict-like argument but with dotted attribute access"?

In the workaround below, the `Deps` `TypedDict` mirrors the `_Deps` `BaseModel`, indicating this is a "leaf" node of the CLI, and a plain old dict becomes a parameter to the leaf command `Binarize`. If I were to place `_Deps` directly in `Binarize` instead, the CLI fails to parse/build.

Of course you might suggest unwrapping `Deps` and placing its args `raw` and `other` directly into `Binarize`, though here I'm implementing a common "pipeline stage" concept with each stage having a certain shape (e.g. all having a `Deps` argument which is a namespace of paths).

I also see the point of keeping dataclass-like classes single-purpose, so if there's no way to make dataclass-like classes act like arguments instead, I would understand the intent behind a design decision like that, especially if undoing that invariant adds overhead to every single class inference made in CLI parsing.

```Python
from __future__ import annotations
from pathlib import Path
from typing import TypedDict

from cappa import command, invoke
from pydantic import BaseModel

class _Deps(BaseModel):
    raw: Path = Path("raw")
    other: Path = Path("other")

class Deps(TypedDict):
    raw: Path
    other: Path

def main(args: Binarize):
    print(args.deps["raw"], args.deps["other"])

@command(invoke=main)
class Binarize(BaseModel):
    deps: Deps = _Deps().model_dump()

invoke(Binarize)
```

**Comment by DanCardin** [comment]
So cappa doesn't automatically infer dataclass annotated things are commands/subcommands, beyond the top-level class being parsed against, i.e. `Binarize` in this case. You need to annotate `Subcommands[Deps]` if you wanted it to be a subcommand.

But i think there are a couple of problems going on here.

1. Cappa is currently mapping against the default value also. e.g. this works:

   ```python
   from __future__ import annotations
   
   from dataclasses import dataclass
   from typing import Annotated
   
   from cappa import Arg, command, invoke
   from pydantic import BaseModel
   
   
   @dataclass
   class Deps:
       raw: str
   
   
   def main(args: Binarize):
       print(args.deps.raw)
   
   
   @command(invoke=main)
   class Binarize(BaseModel):
       deps: Annotated[Deps, Arg(default="foo")]
   
   
   invoke(Binarize)
```

   `binarize.py` yields `foo`

   Whereas with `deps: Deps = Deps("foo")` (which is what you'd want to write, to be type-correct) it yields `Deps(raw='foo')`

   It's probably the case that we shouldn't be performing the type coersion when the default value is selected...? If you agree, that's definitely something I can almost certainly address quickly.

2. Pydantic doesn't allow positional arguments

    The same example above, with pydantic will fail because we perform `Deps(<the-parsed-cli-input>)`, and this will fail pydantic's constructor. I dont see an obvious way around **this**, other than you providing a dedicated `parse` (or by using dataclasses/pydantic.dataclasses, which **does** allow positional arguments) e.g.

   ```python
   class Deps(BaseModel):
       raw: str
       @classmethod
       def new(cls, raw: str):
           return cls(raw)

   ...

   deps: Annotated[Deps, Arg(action=Deps.new)]
   ```

    ...although having said that, i suppose there's a world where we add a default inference for `BaseModel`s, but it sort of leads into my next point

3. It's not clear to me what you expect the CLI shape/behavior to be.

    Both of my above examples simplify your example to a class with one input, which corresponds to that one input on the CLI.

    What are you anticipating the user would input on the CLI and how that would translate into a constructed `Deps` instance?

    For example, one option i could imagine, is introspect the annotated classes' input fields (lets say 2 arguments `raw` and `other`), and from that inferring that that means `num_args=2`, so that the CLI parser for `binarize.py 1 2` gets `("1", "2")`. and then the mapping stage uses the same inference to splat the arguments over the input space `Deps(1, 2)` (or `Deps(raw=1, other=2)` for BaseModel)

   I'm not sure if that's what you're expecting, but i feel like there might be other obvious interpretations and that any automatic choice might be confusing.

I think, depending on your answer to number 3, it should be possible to arrive at the expected CLI shape and behavior, just by explicitly supplying some combination of `num_args` and `parse`. but there may or may not be a sensible automatically inferable option.

**Comment by blakeNaccarato** [comment]
Okay after reading your (very thorough, thanks!) explanation I see what you mean. I've written an example below that operates on an assumption that all bottom-level/leaf commands have arguments `Params`, `Deps`, and `Outs`. I've written an ice-cream maker example where `Deps` are file paths to household appliances (a stretch, I know), `Outs` are file paths where the ice cream will be placed, and `Params` affect ice cream color and flavor. In the proper implementation I'm using this pipeline layout to define different data processes for image processing.

In the example below, I've passed a quick-and-dirty custom backend to `cappa.parse` in order to get the behavior I'm looking for, but in a brittle way.

- I'm trying to get very similar behavior to "Pydantic settings" CLI parsing which can parse `list`/`set`/`dict` values provided as CLI arguments as [JSON-encoded strings](https://docs.pydantic.dev/latest/concepts/pydantic_settings/#parsing-environment-variable-values). A few things I had to do to get Cappa to play along...
- My custom backend first runs the default `cappa.parser.backend` and gets `parser`, `parsed_command`, and `parsed_args`, then further massages those.
- I see how if the parser gets a `BaseModel`/`dataclass` instance, it feeds the instance back to the class constructor. So my custom backend does `arg.parse = arg.parse.model_validate` (blindly, so they all have to be `BaseModel`s), so that later during `map_result`, `value = arg.parse(value)` doesn't explode.
- The parser grabs the default (`cappa.arg.Arg.default`) for every argument (it will fail if there is no default)
- It merges these `defaults` with vanilla `parsed_args` from the standard backend to control behavior downstream (this only works if every argument is in `long=True` form)
- Now  `value = arg.parse(value)` becomes basically `value = my_model_instance.model_validate(value_dict)` and doesn't fail
- Finally `return command.cmd_cls(**kwargs)` at bottom of `Command.map_result` succeeds

The example in the collapsed "details" fold below exercises Cappa in a few different ways, and also shows how Pydantic settings' CLI works with the same `argv`. Now that I've got the behavior correct in a quick-and-dirty way, I think the following things are notable:

- The `backend` parameter of `cappa.base.invoke` is the escape hatch I needed! However, this escape hatch can't be special-cased for certain parts of the CLI, e.g. the `Stage` sub-branch of my CLI which has this `Params`/`Deps`/`Outs` "shape" and behavior. Maybe a given CLI model class could have an optional `backend` on its `@command` decorator that is dispatched for processing that node downward, for instance.
- Is it safe for the user's custom backend to modify `arg.parse` on `cappa.arg.Arg` instances?

I will probably expose this "special" part of my CLI with a bunch of uniform "pipeline stages" from a different entry point so that this wonky custom `backend` doesn't have to handle too much. Imagine a CLI of depth 2 with the root command having each subcommand be a "stage" (and no deeper).

<details>
<summary>Details</summary>

```Python
"""Pipeline stage for making ice cream."""

from __future__ import annotations

import sys
from contextlib import suppress
from json import loads
from pathlib import Path
from typing import Annotated, Any, TypeVar

from cappa.arg import Arg
from cappa.base import command, invoke, parse
from cappa.command import Command
from cappa.output import Output
from cappa.parser import backend
from pydantic import BaseModel
from pydantic_settings import BaseSettings


def main(stage: MakeIceCream | PydanticSettingsMakeIceCream):  # noqa: D103
    print(  # noqa: T201
        f"Here's your {stage.params.color.upper()} ice cream in a {stage.outs.waffle_cone.stem.upper()} waffle cone!\n"
    )


class Params(BaseModel):  # noqa: D101
    color: str = "yellow"
    flavor: str = "vanilla"


class Deps(BaseModel):  # noqa: D101
    freezer: Path = Path("freezer")
    stand_mixer: Path = Path("stand_mixer")


class Outs(BaseModel):  # noqa: D101
    waffle_cone: Path = Path("waffle_cone")
    bowl: Path = Path("bowl")


class MakeIceCream(BaseSettings):
    """Make ice cream."""

    params: Annotated[Params, Arg(long=True)] = Params()
    deps: Annotated[Deps, Arg(long=True)] = Deps()
    outs: Annotated[Outs, Arg(long=True)] = Outs()


class PydanticSettingsMakeIceCream(MakeIceCream, cli_parse_args=True): ...  # noqa: D101


T = TypeVar("T")


def defaults_backend(
    command: Command[T],
    argv: list[str],
    output: Output,
    prog: str,
    provide_completions: bool = False,
) -> tuple[Any, Command[T], dict[str, Any]]:
    """Backend that injects defaults and makes `model_validate` the parse callable."""
    parser, parsed_command, parsed_args = backend(
        command=command,
        argv=argv,
        output=output,
        prog=prog,
        provide_completions=provide_completions,
    )
    extra_args = ["help", "completion"]
    defaults: dict[str, dict[str, Any]] = {}
    for arg in parsed_command.arguments:
        if arg.value_name in extra_args:  # pyright: ignore[reportAttributeAccessIssue]
            continue
        arg.parse = arg.parse.model_validate  # pyright: ignore[reportAttributeAccessIssue, reportFunctionMemberAccess, reportOptionalMemberAccess]
        defaults[arg.value_name] = arg.default.model_dump_json()  # pyright: ignore[reportAttributeAccessIssue, reportOptionalMemberAccess]
    return (
        parser,
        parsed_command,
        {k: loads(v) for k, v in {**defaults, **parsed_args}.items()},  # pyright: ignore[reportArgumentType]
    )


def main2(stage: InvokeMakeIceCream):  # noqa: D103
    print(  # noqa: T201
        f"Here's your {stage.params.color.upper()} ice cream in a {stage.outs.waffle_cone.stem.upper()} waffle cone!\n"
    )


@command(invoke=main2)
class InvokeMakeIceCream(BaseModel):  # noqa: D101
    params: Annotated[Params, Arg(long=True)] = Params()
    deps: Annotated[Deps, Arg(long=True)] = Deps()
    outs: Annotated[Outs, Arg(long=True)] = Outs()


if __name__ == "__main__":
    if len(sys.argv) != 1:
        invoke(InvokeMakeIceCream, backend=defaults_backend)
        sys.exit()
    with suppress(SystemExit):
        print("\nCAPPA HELP:")  # noqa: T201
        main(parse(MakeIceCream, backend=defaults_backend, argv=["--help"]))
    with suppress(SystemExit):
        print("\nCAPPA NO ARGS:")  # noqa: T201
        invoke(InvokeMakeIceCream, backend=defaults_backend)
    sys.argv = [
        "make_ice_cream.py",
        "--params",
        '{"color": "blue"}',
        "--outs",
        '{"waffle_cone": "chocolate_dipped"}',
    ]
    with suppress(SystemExit):
        print("\nCAPPA WITH CUSTOM BACKEND:")  # noqa: T201
        main(parse(MakeIceCream, backend=defaults_backend))
    with suppress(SystemExit):
        print("PYDANTIC SETTINGS:")  # noqa: T201
        main(PydanticSettingsMakeIceCream())
```

</details>

**Reply by DanCardin** [reply, marked as answer]
I'm not entirely sure whether i'm fully grasping the complexities in this example or not. But if it's just a matter of making singular CLI inputs parse into arbitrary shapes, then you ought to be able to use `Arg.parse` to specify the custom mechanism required to turn that string input into the object you want.

I've simplified your example into what seems like the minimal request here.

```python
from __future__ import annotations

from collections.abc import Callable
from json import loads
from typing import Annotated, TypeVar

from cappa.arg import Arg
from cappa.base import command, invoke
from pydantic import BaseModel

T = TypeVar("T", bound=BaseModel)


def main(stage: MakeIceCream):
    print(
        f"Here's your {stage.params.color.upper()} {stage.params.flavor.upper()} ice cream!\n"
    )


class Params(BaseModel):
    color: str = "yellow"
    flavor: str = "vanilla"


def from_json_str(model: type[T]) -> Callable[[str], T]:
    def map(json_str: str):
        value = loads(json_str)
        return model.model_validate(value)

    return map


@command(invoke=main)
class MakeIceCream(BaseModel):
    params: Annotated[Params, Arg(long=True, parse=from_json_str(Params))] = Params().model_dump_json()


if __name__ == "__main__":
    invoke(MakeIceCream)
```

i.e. turn the string into a dict structure, and then `model_validate` it. Or if pydantic had a mechanism for parsing directly from a string, then that instead of `from_json_str`. This also seems to take care of defaults by way of leaving it to pydantic

It **seemed** like the other params, and the use of BaseSettings was mostly irrelevant to this purpose, which is why i omitted them. If that's not the case, lmk.

Also, if it matters, `BaseSettings` also works as a source class.

---

Also note, Cappa ships with a mechanism for loading default values from env vars: https://cappa.readthedocs.io/en/latest/arg.html#environment-variable-fallback, if that would be meaningful to you. Also I happened to write https://github.com/DanCardin/dataclass-settings/ which (imo) fixes a bunch of problematic things in pydantic-settings, which i just know definitely works with cappa.

---

Also like i said in my o.g. response, i think the fact that we're having to work around the default being required to be a string, when the annotation says otherwise is definitely what i'd consider a bug.

**Reply by blakeNaccarato** [reply]
I used `backend` after seeing it in the documentation, not realizing it was a hacksaw to the scalpel of `Arg(parse=...)`. Obviously "rtfm" may apply to me here 😅. A few thoughts in conclusion here:

- It might help to redirect wayward travelers to the "Parser Backends" page with a "you may be looking for `Arg(parse=...)`" note.
- I see `Arg(parse=...)` makes guest appearances in Apps/Annotations "Concepts" pages and API docs. I see the difficulty in separating concepts here, since "parsing" is an overloaded term that means overall behavior as well as detailed, field-specific `Arg(parse=...)` behavior.
- Maybe a good "Parsing" Concept article would acknowledge this tension, written in such a way as to discuss the defaults, then escalate from minimally-invasive parsing customization to full-blown `backend` overrides. This might take care of some of the responsibilities of the "Args" and "Annotations" pages, allowing them to become more focused as well.
- Someday I swear I'll graduate and actually get some more time to contribute documentation and upstream changes to the packages I love, and that _someday_ is when I'd be open to helping with this!

I also appreciate that you have in your back pocket an entire "`pydantic-settings` but they did it wrong" package ready to go. I recently got intimately familiar with `pydantic-settings` in implementing custom `settings_customize_sources` for certain behavior, so I understand the pain there. Though `pydantic-settings` is definitely helpful when you want custom class initialization behavior without hijacking `__init__`. Will be interesting to see if Pydantic plugins, once stabilized, provide a cleaner alternative to that.

Anyways, here's my final example for making ice cream in the theme of a pipeline stage (an odd example, I know). Thanks again for your help here!

```Python
"""Pipeline stage for making ice cream."""

from __future__ import annotations

from collections.abc import Callable
from json import loads
from pathlib import Path
from sys import argv
from typing import Annotated, TypeVar

from cappa.arg import Arg
from cappa.base import command, invoke
from pydantic import BaseModel


def main(stage: MakeIceCream):
    print(
        f"Here's your {stage.params.color.upper()} ice cream in a {stage.outs.waffle_cone.stem.upper()} waffle cone!\n"
    )


class Params(BaseModel):
    color: str = "default_yellow"
    flavor: str = "default_vanilla"


class Deps(BaseModel):
    freezer: Path = Path("default_freezer")
    stand_mixer: Path = Path("default_stand_mixer")


class Outs(BaseModel):
    waffle_cone: Path = Path("default_cone")
    bowl: Path = Path("default_bowl")


T = TypeVar("T", bound=BaseModel)


def get_parser(model: type[T]) -> Callable[[str], T]:
    """Get parser for model or JSON-encoded string."""

    def parse(v: T | str) -> T:
        """Parse model or JSON-encoded string."""
        return model(**loads(v)) if isinstance(v, str) else v

    return parse


@command(invoke=main, default_long=True)
class MakeIceCream(BaseModel):
    """Make ice cream."""

    params: Annotated[Params, Arg(parse=get_parser(Params))] = Params()
    deps: Annotated[Deps, Arg(parse=get_parser(Deps))] = Deps()
    outs: Annotated[Outs, Arg(parse=get_parser(Outs))] = Outs()


if __name__ == "__main__":
    if len(argv) > 1:
        invoke(MakeIceCream)
    else:
        invoke(
            MakeIceCream,
            argv=[
                "--outs",
                '{"waffle_cone": "chocolate_dipped"}',
                "--params",
                '{"color": "blue"}',
            ],
        )
```

**Reply by DanCardin** [reply]
> I also appreciate that you have in your back pocket an entire "pydantic-settings but they did it wrong" package ready to go. I recently got intimately familiar with pydantic-settings in implementing custom settings_customize_sources for certain behavior, so I understand the pain there. Though pydantic-settings is definitely helpful when you want custom class initialization behavior without hijacking __init__. Will be interesting to see if Pydantic plugins, once stabilized, provide a cleaner alternative to that.

heh. I really wanted it to work after developing cappa, but it kept being problematic trying to integrate it into clis and having no idea how to backtrack to which specific env vars were failing to load.

There are inherent limitations to how they designed it that imo leave the whole thing in a weird place. And it seemed really pompous of me to go to their issue tracker and say: "hey, i like the library...but it should really be designed this way". so alas a new library need be born instead 🤷. And for my own purposes, the extra parsing logic they do is not what I'm looking for. I dont need to harp on it here, though, I lay it all out in the readme 😬.

With that said, I think the CLI parsing stuff you're talking about didn't exist the last time I looked.

**Comment by blakeNaccarato** [comment]
And to avoid burying the lede in that other reply, I'll quote the other actionable bit here that you have mentioned...

> `binarize.py` yields `foo`. Whereas with `deps: Deps = Deps("foo")` (which is what you'd want to write, to be type-correct) it yields `Deps(raw='foo')`. It's probably the case that we shouldn't be performing the type coersion when the default value is selected...? If you agree, that's definitely something I can almost certainly address quickly.

And...

> ...although having said that, i suppose there's a world where we add a default inference for `BaseModel`s, but it sort of leads into my next point
> For example, one option i could imagine, is introspect the annotated classes' input fields (lets say 2 arguments raw and other), and from that inferring that that means num_args=2, so that the CLI parser for binarize.py 1 2 gets ("1", "2"). and then the mapping stage uses the same inference to splat the arguments over the input space Deps(1, 2) (or Deps(raw=1, other=2) for BaseModel)

Which could make sense. The `pydantic_settings` CLI parsing sees a `BaseModel` like `Deps` and allows dotted unpacking like `--deps.raw something --deps.other 'something else'`, but maybe that's too opinionated. Or allowing another expansion point in `backend` that's a bit less of a hacksaw, and by default it does the `--deps.raw something --deps.other 'something else'` thing, but can be customized (per-field? idk). It gets messy as you add more places for users to hijack the arg consumption logic.

And...

> Also like i said in my o.g. response, i think the fact that we're having to work around the default being required to be a string, when the annotation says otherwise is definitely what i'd consider a bug.

Which I now think I understand after digging through the internals, which is that the structure of `cappa.command.Command.map_result` uses the result of `cappa.parser.backend` first, which has its own `cappa.parser.parse` to consume the context, ultimately relying on `cappa.parser.consume_arg`, which is where my comprehension falls apart a bit.

But the crux is that in `cappa.command.Command.map_result`, by the time we get to the `# apply Arg(parse=...) or else default parsing` comment, it's already too late to have handled the string CLI input differently, right? And so it's good we have `Arg(parse=...)`, but you're suggesting that some rethinking of arg consumption prior to that is possible? Here's the trimmed down `cappa.command.Command.map_result` I'm referencing:

```Python
def map_result(self, command: Command[T], prog: str, parsed_args) -> T:
    kwargs = {}
    for arg in self.value_arguments():
        is_subcommand = isinstance(arg, Subcommand)  # Side note: this var could be used below...
        if arg.field_name not in parsed_args:
            # handle default
        else:
            # get arg as parsed by the backend
        if ...:
            # handle other special stuff
        if isinstance(arg, Subcommand):  # Side note: This could be `is_subcommand` I think?
            # handle subcommand
        else:
            # apply `Arg(parse=...)` or else default parsing
        kwargs[arg.field_name] = value
    return command.cmd_cls(**kwargs)
```

**Reply by DanCardin** [reply]
Lot to respond to :P 

I think my steps, in order, as a result of this are probably:

* Rework how `Arg.parse` is handled, such that the default is given back, unmapped.
  * Perhaps either RuntimeError if it doesn't match type-wise, or else perhaps warn during CLI setup if the default doesn't match the type
* Consider automatically calling `model_validate` on behalf of a BaseModel
  * Probably by also adding a `parse_json` "combinator" you can just pass to `parse` without having to define yourself.
* docs
  * to backend for clarification
  * to parse/action specifically probably
  * I have in-progress stuff more generally for examples/comparisons, but that's still taking form.

---
> The pydantic_settings CLI parsing sees a BaseModel like Deps and allows dotted unpacking like --deps.raw something --deps.other 'something else', but maybe that's too opinionated.

I'm definitely not going to automatically infer `--foo.bar` arguments. What I could imagine myself doing is shipping a custom `Arg.action` where you'd define something like:

```python
foo: Annotated[Params, DestructuredArg(long=True)]
```

which would be equivalent to something like
```python
def restructure(context: ParseContext, arg: Arg, value: Value):
    pre, post = arg.field_name.split('.')
    result = context.result.setdefault(pre, {})
    result[post] = value.value
    return result

foo: Annotated[Params, Arg(field_name='foo.color', long=True, action=restructure), Arg(field_name='foo.flavor', long=True, action=restructure)]
```

so that `--foo.color` and `--foo.flavor` are parsed into the same dict structure, and the reconstructed out the other end as part of the same thing. I could imagine this being at least generally useful, if not something i'd want as default behavior...if this does in fact work....

**Reply by blakeNaccarato** [reply]
> Rework how `Arg.parse` is handled, such that the default is given back, unmapped. [...] Perhaps either RuntimeError if it doesn't match type-wise, or else perhaps warn during CLI setup if the default doesn't match the type

Ah, yeah that makes sense. Probably similar concerns led to the creation of Pydantic's `validate_default` model/field config, but the "magic secret config option" approach is less nice for Cappa, which is trying to be a thin orchestrator around generic dataclass-like interfaces. So it's probably good to choose one (e.g. "We're not going to map default values, no questions asked!"), and let users who _really need_ alternative behavior supply their own `backend`.

I think that generally the current "shape" of parsing is very sensible, and delegating to the field-level `Arg` config is a good pattern, even if a little verbose. So it makes sense to confine changes to `Arg` semantics rather than going into the `backend` and its call chain.

> Consider automatically calling `model_validate` on behalf of a BaseModel.

Sounds good, though I understand the point of parsing solely within Cappa and avoiding relying on any special behavior of `BaseModel`s in the Cappa parsing flow.

> Probably by also adding a `parse_json` "combinator" you can just pass to `parse` without having to define yourself.

Yeah, I think facilitating parsing of JSON-encoded strings is a sensible toggle for structures passed via CLI. It might make sense to have that configurable as a CLI-wide thing via `@command` like how `default_long` is handled. Though I know it starts with a few options today, and before you know it `@command` starts growing defaults for everything `Arg` is meant to do, and it gets messy. I like the limited scope of `default_long`, that it doesn't propagate down through subcommands. Explicit is better than implicit and all that jazz.

> docs, to backend for clarification, to parse/action specifically probably, I have in-progress stuff more generally for examples/comparisons, but that's still taking form.

Ah, good to know.

> I'm definitely not going to automatically infer `--foo.bar` arguments. What I could imagine myself doing is shipping a custom `Arg.action` [...] so that `--foo.color` and `--foo.flavor` are parsed into the same dict structure, and the reconstructed out the other end as part of the same thing. I could imagine this being at least generally useful, if not something i'd want as default behavior...if this does in fact work....

Yeah, I figured that's definitely too magical, and would have a compounding effect on the number of possible representations for deeper hierarchies!

Anyways, this was a helpful back-and-forth for me, and I definitely feel more comfortable with Cappa's semantics now that I've had a crash course in the backend. Thanks!

**Reply by DanCardin** [reply]
Fix for item 1 so far: https://github.com/DanCardin/cappa/pull/138

**Reply by DanCardin** [reply]
I think https://github.com/DanCardin/cappa/pull/140 addresses some of the docs issues. i still have an in-progress bit of docs on more general examples things.

And i think https://github.com/DanCardin/cappa/pull/141 will eventually be "the solution" to item 2. although it's 1. not automatic (for reasons stated in the pr), and it's blocked on adopting type-lens.

**Reply by DanCardin** [reply]
actually...i guess pydantic BaseModel classes **could** perhaps be automatically inferred to use `model_validate` (and you'd still need to supply `parse=json.loads`) because the `__init__` args are required to be kwargs, which will never work by default (without something like #141).

**Reply by DanCardin** [reply]
...and the last thing i need to test out is my `DestructuredArg` idea. that...should essentially boil down to the same output shape as a `parse=json.loads` (i.e. a dict), but with disparate args; but would then need equivalent handling visa-vi these previous comments ^

**Reply by blakeNaccarato** [reply]
The docs changes address confusion about parsing very nicely, I like the mention of `parse_` commands and cross-linking between confusable concepts.

I think it definitely makes sense to solve dataclass inference by just waiting a year for 3.8 to fully drop off and `type-lens` to progress. Like it's not an outright blocker since the "wrap the backend" escape hatch is there and not too painful. I agree that baking in ugly inference today is less attractive if `type-lens` facilitates that nicely in the future. I've manually parsed my share of `BaseModel` annotations and it gets ugly pretty quickly with all the `get_args`, `get_origin`, `is_subclass` boilerplate.

I think deferring to `BaseModel.model_validate` could work, though if built-in dataclass-like inference is easier to implement in a year or so, that would be less brittle and not Pydantic-specific. So it's less churn to just not touch it now.

Implementing the idea of destructured args sounds promising, that seems the right-shaped idea for handling sub-`Arg` specifics. The caveat being that the atomicity of an `Arg` is a nice conceptual "floor" to the already very complex logic of commands.

With `Args` being structurally similar to subcommands already (they're kind of like always-leaf combinatorial subcommands in a way), maybe there's an implementation that reuses some of that machinery and minimizes the impedance mismatch when moving from `Subcommand` land to `Arg` land. But I have less exposure to the concepts so maybe the confusion lies with my inexperience.

Thanks for the thorough consideration of these ideas!

**Reply by DanCardin** [reply]
> I think it definitely makes sense...

fwiw, i think type-lens is an option soon, i'm just trying to merge interfaces they're currently lacking. they backported it to 3.8 recently (although unreleased).

it's mainly that i found myself needing to expose the annotation shape as a user accessible feature, which i dont want to be the current raw type shape. (currently **you** can only make inferences off the input value, but **cappa** knows the annotated type also, so making the decision to infer dataclass-likes a user's `parse=` choice then becomes a problem, unless they're forced to duplicate the annotation into the parse call like `Annotated[Foo, Arg(parse=as_dataclass(Foo))]`, which is sad).

---

> I think deferring to BaseModel.model_validate...

The nice thing about `model_validate` is that is just accepts a single-argument blob, which means a dict (`json.loads`) works fine (but also apparently you can `pydantic.TypeAdapter(Foo).validate_json` fwiw), whereas the constructor requires `**splat`ing a dict.

---

> Implementing the idea of destructured args sounds promising,...

As far as destructured args go, one can **already** do `foo: Annotated[dict, Arg(long='--foo', action=something_fancy), Arg(long='--bar', action=something_fancy)]` to produce 2 cli-level arguments that target the same `field_name`. in this case `something_fancy` would presumably write them to dedicated dict keys.

the fancy bit would "just" be translating a dataclass-like into the appropriate series of `Arg`s.

---

> With Args being structurally similar to subcommands already...

I'm not sure i follow what you mean here. unless you mean like... `foo.py bar`, where `bar` could either be a subcommand or a positional argument? if so, the difference is that subcommands move you into a completely different parser context, whereas args do not (and thus you can `--foo=bar bar --baz=baz` intermingle "options" around the arg). so i'm not sure what to make of this comment either way :P 

---

> Thanks for the thorough consideration of these ideas!

Thank **you**! Other than discovering bugs or inference extensions, these sort of ideas are the main way the capabilities get improved!

**Reply by pawamoy** [reply]
**cappa**bilities :grinning:

**Reply by blakeNaccarato** [reply]
Regarding my conflating subcommand and destructured arg shapes, chalk it up to my flawed mental model. Okay I'll stop trading essays, you're free to close the discussion thread, if/when that's part of your workflow!

Thanks @pawamoy for the Cappa mention on Mastodon! It turned out to be the missing piece for representing [my pipeline stages](https://github.com/softboiler/boilercv/blob/main/pipeline%2Fboilercv_pipeline%2Fstages%2Ffind_contours%2F__init__.py) suitably as [commands](https://dvc.org/doc/user-guide/project-structure/dvcyaml-files#stage-commands) in a pipeline orchestrator's [`dvc.yaml`](https://github.com/softboiler/boilercv/blob/main/dvc.yaml#L39-L47) (combined with a little [YAML dumping](https://github.com/softboiler/boilercv/blob/9a1d70975ebaba43fe29765721ba5edadcb31a19/pipeline/boilercv_pipeline/cli.py#L65)). It sure beats the alternative, manually syncing `dvc.yaml` with hardcoded params/deps/outs. With Cappa you also get a human-usable CLI out of it, to boot!

**Comment by DanCardin** [comment]
https://github.com/DanCardin/cappa/pull/150: Merged
https://github.com/DanCardin/cappa/pull/141: still cooking but promising. Although it makes the whole parsing apparatus decently more complicated. i'd want to make sure it doesn't have negative downstream effects.

**Comment by blakeNaccarato** [comment]
The argument destructuring concept looks great! In the interim I flattened a bunch of [format-related args](https://github.com/softboiler/boilercv/blob/b09ccdc28a3cd097b803ee32b89c89f1b004ee30/packages/pipeline/boilercv_pipeline/models/params/__init__.py#L116) from a `Format` class into the parent `Params` class used as a `cappa.Subcommand`. Now I can afford to isolate those format-specific things back into their own `Format` class and use arg destructuring to assemble them.

Handling arg destructuring as a simple extension of parent args during [command collection](https://github.com/DanCardin/cappa/blob/49cd4838745de4d2aff572495842b8bc4f0a6d6a/src/cappa/arg.py#L208) is a great way to handle this. Seems a cleaner approach, rather than expecting the destructured arg name be pre-pended e.g. `--sub-object-arg-1 foo` in the way of `pydantic-settings`'s CLI feature mentioned earlier.

In the [docs example](https://cappa.readthedocs.io/en/latest/arg.html#arg-destructured-arg-destructure), if `Args` later grows an `arg1` field itself, will that be detected as a collision with `SubObject`'s `arg1` field? Or will one silently take precedence?

**Reply by DanCardin** [reply]
oo good question. So if they're both options, then it'll fail in the parser with `f"Conflicting option string: {key}"`.

but if they otherwise dont conflict at the CLI-shape level (e.g. 2 positionals, or one positional and one option), then it **should** "just work" because they're recorded to different underlying `field_name`s (`sub1` vs `field.sub1`).

With that said, i'm glad you thought of that because there's currently a bug where it will record the value for both fields, easily remedied (https://github.com/DanCardin/cappa/pull/151)

**Reply by blakeNaccarato** [reply]
Ah, glad I could help! Well it's a great feature and I will be making use of it sometime in the near future.
